pkg/agent/nplagent/k8s/annotations.go

```diff
@@ -0,0 +1,164 @@
+	NodePort string `json:"Nodeport"`
+}
+
+func Stringify(serialize interface{}) string {
+	json_marshalled, _ := json.Marshal(serialize)
+	return string(json_marshalled)
+}
+
+func IsNodePortInAnnotation(s []NPLEPAnnotation, nodeport string) bool {
+	for _, i := range s {
+		if i.NodePort == nodeport {
+			return true
+		}
+	}
+	return false
+}
+
+func assignPodAnnotation(pod *corev1.Pod, containerPort, nodeIP, nodePort string) {
+	var err error
+	current := make(map[string]string)
+	if pod.Annotations != nil {
+		current = pod.Annotations
+	}
+
+	klog.Infof("Building annotation for pod: %s\tport: %s --> %s:%s", pod.Name, containerPort, nodeIP, nodePort)
+
+	var annotations []NPLEPAnnotation
+	// nplEP annotation exists
+	if current[NPLAnnotationStr] != "" {
+		if err = json.Unmarshal([]byte(current[NPLAnnotationStr]), &annotations); err != nil {
+			klog.Warningf("Unable to unmarshal NPLEP annotation")
+		}
+
+		if !IsNodePortInAnnotation(annotations, nodePort) {
+			annotations = append(annotations, NPLEPAnnotation{
+				PodPort:  containerPort,
+				NodeIP:   nodeIP,
+				NodePort: nodePort,
+			})
+		} else {
+			// mapping for the containerPort already exists
+			// TODO
+		}
+	} else {
+		annotations = []NPLEPAnnotation{NPLEPAnnotation{
+			PodPort:  containerPort,
+			NodeIP:   nodeIP,
+			NodePort: nodePort,
+		}}
+	}
+
+	current[NPLAnnotationStr] = Stringify(annotations)
+	pod.Annotations = current
+}
+
+func removeFromPodAnnotation(pod *corev1.Pod, containerPort string) {
+	var err error
+	current := pod.Annotations
+
+	klog.Infof("Removing annotation from pod: %s\tport: %s", pod.Name, containerPort)
+	var annotations []NPLEPAnnotation
+	if err = json.Unmarshal([]byte(current[NPLAnnotationStr]), &annotations); err != nil {
+		klog.Warningf("Unable to unmarshal NPLEP annotation")
+		return
+	}
+
+	for i, ann := range annotations {
+		if ann.PodPort == containerPort {
+			annotations = append(annotations[:i], annotations[i+1:]...)
+			break
+		}
+	}
+
+	current[NPLAnnotationStr] = Stringify(annotations)
+	pod.Annotations = current
+}
+
+// RemoveNPLAnnotationFromPods : Removes npl annotations from all pods
+func (c *Controller) RemoveNPLAnnotationFromPods() {
+	podList, err := c.KubeClient.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		klog.Warningf("Unable to list Pods")
+		return
+	}
+	nodeName, err := env.GetNodeName()
+	if err != nil {
+		klog.Warningf("Failed to get nodename, NPL annotation can not be removed for pods")
+		return
+	}
+	for _, pod := range podList.Items {
+		if nodeName != pod.Spec.NodeName {
+			continue
+		}
+		podAnnotation := pod.GetAnnotations()
+		if podAnnotation == nil {
+			continue
+		}
+		klog.Infof("Removing all NPL annotation from pod: %s, ns: %s", pod.Name, pod.Namespace)
+		delete(podAnnotation, NPLAnnotationStr)
+		pod.Annotations = podAnnotation
+		c.KubeClient.CoreV1().Pods(pod.Namespace).Update(context.TODO(), &pod, metav1.UpdateOptions{})
```

Comment:
Will there be problem if the update fails?

Reply:
There will be a problem in that Case. I'll add a log here.